modules/rpc/yarpc.go

```diff
@@ -105,78 +176,105 @@ func newYarpcModule(
+
+	tchTransport, err := tch.NewChannelTransport(
+		tch.ServiceName(module.config.AdvertiseName),
+		tch.ListenAddr(module.config.Bind),
+	)
 
-	module.inboundMiddlewares = inboundMiddlewaresFromCreateInfo(mi)
-	module.onewayInboundMiddlewares = onewayInboundMiddlewaresFromCreateInfo(mi)
+	module.config.inbounds = []transport.Inbound{tchTransport.NewInbound()}
 
 	return module, err
 }
 
-// Start begins serving requests over YARPC
-func (m *YarpcModule) Start(readyCh chan<- struct{}) <-chan error {
+// Start begins serving requests over RPC. It first stops the current dispatcher, merges configs from the global
+// collection and then starts the new dispatcher.
+func (m *YARPCModule) Start(readyCh chan<- struct{}) <-chan error {
+	if m.IsRunning() {
+		panic("module is already running")
+	}
+
 	m.stateMu.Lock()
 	defer m.stateMu.Unlock()
 
-	interceptor := yarpc.UnaryInboundMiddleware(m.inboundMiddlewares...)
-	onewayInterceptor := yarpc.OnewayInboundMiddleware(m.onewayInboundMiddlewares...)
-
 	ret := make(chan error, 1)
-	tchTransport, err := tch.NewChannelTransport(
-		tch.ServiceName(m.config.AdvertiseName),
-		tch.ListenAddr(m.config.Bind),
-	)
-	if err != nil {
-		ret <- errors.New("Unable to create TChannel transport " + err.Error())
+	if err := _configs.addConfig(&m.config); err != nil {
+		ret <- errors.New("can't add module config " + err.Error())
 		return ret
 	}
 
-	m.rpc, err = _dispatcherFn(m.Host(), yarpc.Config{
-		Name: m.config.AdvertiseName,
-		Inbounds: []transport.Inbound{
-			tchTransport.NewInbound(),
-		},
-		InboundMiddleware: yarpc.InboundMiddleware{
-			Unary:  interceptor,
-			Oneway: onewayInterceptor,
-		},
-		Tracer: m.Tracer(),
-	})
+	config, err := _configs.mergeConfigs()
+
+	if err != nil {
+		ret <- errors.New("unable to merge configs from multiple modules " + err.Error())
+		return ret
+	}
 
+	dispatcher, err := _dispatcherFn(m.Host(), config)
 	if err != nil {
 		ret <- err
 		return ret
 	}
 
+	if err := resetDispatcher(dispatcher); err != nil {
+		ret <- err
+		return ret
+	}
+
 	m.register(m)
-	// TODO update log object to be accessed via context.Context #74
-	m.log.Info("Service started", "service", m.config.AdvertiseName, "port", m.config.Bind)
+	m.Host().Logger().Info("Module started",
+		"name", m.config.AdvertiseName,
+		"port", m.config.Bind)
 
-	ret <- m.rpc.Start()
+	m.isRunning = true
 	readyCh <- struct{}{}
 	return ret
 }
 
 // Stop shuts down a YARPC module
-func (m *YarpcModule) Stop() error {
+func (m *YARPCModule) Stop() error {
+	if !m.IsRunning() {
+		panic("module is not running")
+	}
+
 	m.stateMu.Lock()
 	defer m.stateMu.Unlock()
 
-	if m.rpc != nil {
-		err := m.rpc.Stop()
-		m.rpc = nil
+	if err := _configs.removeConfig(&m.config); err != nil {
+		return err
+	}
+
+	config, err := _configs.mergeConfigs()
+
+	if err != nil {
+		return errors.New("unable to merge configs from multiple modules " + err.Error())
+	}
+
+	dispatcher, err := _dispatcherFn(m.Host(), config)
```

Comment:
seems to be an almost copy-paste of above (call _dispatcherFn, resetDispatcher).